### PR TITLE
chore(Authoring): Convert project asset authoring to Angular

### LINKED ISSUE #1337: chore(Authoring): Convert ProjectAssetAuthoringComponent to Angular
Convert projectAssetAuthoringComponent.ts.

### PR DESCRIPTION
**Pull request by geoffreykwan**
## Changes

Converted project asset authoring to Angular.
Note: We will delete ProjectAssetService.openAssetChooser() after we convert ProjectInfoAuthoringController to Angular because that view still uses the old asset chooser.

## Test

- Make sure the asset authoring view still works properly
- Make sure the select asset popup still works properly (like when choosing an image in various places like draw background image, match choice, multiple choice choice, html component insert asset, etc.)
- Make sure drag and drop as well as click to open file chooser both work
- Make sure you can only upload accepted file types

Closes #1337

### CODE REVIEW

File: src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+  constructor(
+    private configService: ConfigService,
+    @Optional() @Inject(MAT_DIALOG_DATA) private dialogData: any,
+    @Optional() protected dialogRef: MatDialogRef<ProjectAssetAuthoringComponent>,
+    private projectAssetService: ProjectAssetService
+  ) {}
+
+  ngOnInit(): void {
+    if (this.dialogData != null) {
```
Comment (hirokiterashima) [inline src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts:47]: Is this null-check necessary?

File: src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+  targetObject = null;
+  totalFileSize = 0;
+  totalUnusedFilesSize = 0;
+  unusedFilesPercentage = 0;
+  uploadSuccessMessage = '';
+  uploadErrorMessage: any;
+
+  constructor(
+    private configService: ConfigService,
+    @Optional() @Inject(MAT_DIALOG_DATA) private dialogData: any,
```
Comment (hirokiterashima) [inline src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts]: Create ProjectAssetData to represent fields (and initial values) that are contained in the data to pass into this dialog?

File: src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        this.setTotalUnusedFilesSize(totalUnusedFilesSize);
+      })
+    );
+
+    if (this.projectAssetService.isProjectAssetsAvailable()) {
+      this.projectAssetService.calculateAssetUsage();
+    }
+  }
+
+  private setParams(params: any): void {
```
Comment (hirokiterashima) [inline src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts:78]: Do we need these fields [isPopup/nodeId/componentId], or can they be directly obtained from this.dialogData.[isPopup/nodeId/componentId...]?

**Review by hirokiterashima — CHANGES_REQUESTED** [review]
Functionality seems to work well. I have few comments/suggestions.

1. Change the mouse to pointing hand image when hovering over area to click/drag assets?
2. This change set will introduce a lot of code duplication and worsen the overall code quality (35 new issues). Can you try to reduce it? Is ProjectAssetAuthoringController still needed? 
3. See inline code comments.

File: src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts (hunk 1 above)
Comment (geoffreykwan) [reply src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts:47]: dialogData is null when the author goes to the asset view like http://localhost:81/teacher/edit/unit/54/asset

File: src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts (hunk 2 above)
Comment (geoffreykwan) [reply src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts]: Done

File: src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts (hunk 3 above)
Comment (geoffreykwan) [reply src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts:78]: What happens when dialogData is null?

**Comment by geoffreykwan** [comment]
We will delete projectAssetAuthoringComponent.ts after we convert ProjectInfoAuthoringController to Angular because that view still uses the old asset chooser.

File: src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+  allowedFileTypes: string[] = ['any'];
+  assetIsImage: boolean;
+  assetIsVideo: boolean;
+  assetSortBy = 'aToZ';
+  componentId = null;
+  errorFiles: any;
+  isPopup = false;
+  nodeId = null;
+  previewAssetURL: string;
+  projectId: number;
```
Comment (hirokiterashima) [inline src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts]: It doesn't look like projectId is ever set/used. Delete?

File: src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+  }
+
+  protected downloadAsset(assetItem: any): void {
+    this.projectAssetService.downloadAssetItem(assetItem);
+  }
+
+  protected chooseAsset(assetItem: any): void {
+    const params = {
+      assetItem: assetItem,
+      projectId: this.projectId,
```
Comment (hirokiterashima) [inline src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts]: It doesn't look like projectId is ever set/used. Delete?

File: src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+  private sortAssetsAToZ(a: any, b: any): number {
+    const aFileName = a.fileName.toLowerCase();
+    const bFileName = b.fileName.toLowerCase();
+    if (aFileName < bFileName) {
+      return -1;
+    } else if (aFileName > bFileName) {
+      return 1;
+    }
+    return 0;
```
Comment (hirokiterashima) [inline src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts]: use localeCompare() to reduce complexity?

File: src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+  private sortAssetsSmallToLarge(a: any, b: any): number {
+    const aFileSize = a.fileSize;
+    const bFileSize = b.fileSize;
+    if (aFileSize < bFileSize) {
+      return -1;
+    } else if (aFileSize > bFileSize) {
+      return 1;
+    }
+    return 0;
```
Comment (hirokiterashima) [inline src/assets/wise5/authoringTool/project-asset-authoring/project-asset-authoring.component.ts]: Use subtraction to reduce complexity?

**Review by hirokiterashima — APPROVED** [review]
Ok, new changes look good and I understand your comments. I added few more minor code change suggestions/questions that you can address before you merge.

**Pull request merged**